[FX] Mutable Graph APIs

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45775 [FX] Track use nodes in Node
* **#45227 [FX] Mutable Graph APIs**

Introduced several `Graph` APIs:

`Graph.move_node_before(self, to_move : Node, before : Node)` -> Move node `to_move` before `before` in the Graph. Both `Node` arguments must be present in this graph.

`Graph.erase_node(self, to_erase : Node)` -> Erases the node `to_erase` from the `Graph`. Throws an exception if there are still uses of that node in the `Graph`.

`Node.find_uses(self) -> List['Node']` -> Find all nodes that use the value produced by `self`. The complexity of this function is linear in the number of nodes * number of arguments to each node.
Note that len(find_uses()) is not necessarily equal to attribute `uses`. This node could be used multiple times in the same `Node`. In that case, the user node would appear once in the return value here, but `uses` would account for the total number of times this Node is used by the user node. e.g. a node for `x + x` would have two uses for the `x` node, but the `x + x` node would appear once in the return from `find_uses`

`Node.replace_all_uses_with(self, replace_with : 'Node') -> List['Node']` -> Replace all uses of `self` in the Graph with the Node `replace_with`. Returns the list of nodes on which this change was made.

Also adds a context manager, `torch.fx.graph.insert_before` that can be used to make node insertion happen *before* the passed-in node. For example:

```
graph = torch.fx.Graph()
x = torch.fx.Proxy(graph.placeholder('x'))
relu = torch.relu(x)

with torch.fx.graph.insert_before(relu.node):
    y = torch.neg(x)
    z = torch.tanh(y)

graph.output((relu.node, z.node))
```

Results in graph

```
graph(x):
    %neg_1 : [uses=1] = call_function[target=<built-in method neg of type object at 0x7f6d19a3d6a0>](args = (%x,), kwargs = {})
    %tanh_1 : [uses=2] = call_function[target=<built-in method tanh of type object at 0x7f6d19a3d6a0>](args = (%neg_1,), kwargs = {})
    %relu_1 : [uses=2] = call_function[target=<built-in method relu of type object at 0x7f6d19a3d6a0>](args = (%x,), kwargs = {})
    return (relu_1, tanh_1)
```

Differential Revision: [D23880730](https://our.internmc.facebook.com/intern/diff/D23880730)

Commit:
[FX][WIP] Mutable Graph APIs

[ghstack-poisoned]

Commit:
Update on "[FX][WIP] Mutable Graph APIs"


Introduced several `Graph` APIs:

`insert_node_before(self, n : Node, before : Node) -> Node:` -> insert node `n` directly before `before` in the topological ordering of nodes

`replace_all_uses_with(self, to_replace : Node, replace_with : Node)` -> Find all uses of the value produced by `to-replace` and swap out those references for the value produced by `replace_with`. Does not delete `to_replace`.

`erase_node(self, to_erase : Node)` -> Erases the node `to_erase` from the `Graph`. Throws an exception if there are still uses of that node in the graph.


Differential Revision: [D23880730](https://our.internmc.facebook.com/intern/diff/D23880730)

[ghstack-poisoned]

Commit:
Update on "[FX][WIP] Mutable Graph APIs"


Introduced several `Graph` APIs:

`insert_node_before(self, n : Node, before : Node) -> Node:` -> insert node `n` directly before `before` in the topological ordering of nodes

`replace_all_uses_with(self, to_replace : Node, replace_with : Node)` -> Find all uses of the value produced by `to-replace` and swap out those references for the value produced by `replace_with`. Does not delete `to_replace`.

`erase_node(self, to_erase : Node)` -> Erases the node `to_erase` from the `Graph`. Throws an exception if there are still uses of that node in the graph.

`find_all_uses_of(self, to_find : Node) -> List[Node]` -> Find all nodes that use `to_find` in their args or kwargs

Also add a context manager, `torch.fx.graph.WithInsertaionPoint` that can be used to make node insertion happen *before* the passed-in node. For example:

```
        graph = torch.fx.Graph()
        x = torch.fx.Proxy(graph.placeholder('x'))
        relu = torch.relu(x)

        with torch.fx.graph.WithInsertPoint(relu.node):
            y = torch.neg(x)
            z = torch.tanh(y)

        graph.output((relu.node, z.node))
```

Results in graph

```
graph(x):
    %neg_1 : [uses=1] = call_function[target=<built-in method neg of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    %tanh_1 : [uses=1] = call_function[target=<built-in method tanh of type object at 0x7fa87adb1500>](args = (%neg_1,), kwargs = {})
    %relu_1 : [uses=1] = call_function[target=<built-in method relu of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    return (%relu_1, %tanh_1)
```

Differential Revision: [D23880730](https://our.internmc.facebook.com/intern/diff/D23880730)

[ghstack-poisoned]

Commit:
Update on "[FX][WIP] Mutable Graph APIs"


Introduced several `Graph` APIs:

`insert_node_before(self, n : Node, before : Node) -> Node:` -> insert node `n` directly before `before` in the topological ordering of nodes

`replace_all_uses_with(self, to_replace : Node, replace_with : Node)` -> Find all uses of the value produced by `to-replace` and swap out those references for the value produced by `replace_with`. Does not delete `to_replace`.

`erase_node(self, to_erase : Node)` -> Erases the node `to_erase` from the `Graph`. Throws an exception if there are still uses of that node in the graph.

`find_all_uses_of(self, to_find : Node) -> List[Node]` -> Find all nodes that use `to_find` in their args or kwargs

Also adds a context manager, `torch.fx.graph.WithInsertionPoint` that can be used to make node insertion happen *before* the passed-in node. For example:

```
graph = torch.fx.Graph()
x = torch.fx.Proxy(graph.placeholder('x'))
relu = torch.relu(x)

with torch.fx.graph.WithInsertPoint(relu.node):
    y = torch.neg(x)
    z = torch.tanh(y)

graph.output((relu.node, z.node))
```

Results in graph

```
graph(x):
    %neg_1 : [uses=1] = call_function[target=<built-in method neg of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    %tanh_1 : [uses=1] = call_function[target=<built-in method tanh of type object at 0x7fa87adb1500>](args = (%neg_1,), kwargs = {})
    %relu_1 : [uses=1] = call_function[target=<built-in method relu of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    return (%relu_1, %tanh_1)
```

Differential Revision: [D23880730](https://our.internmc.facebook.com/intern/diff/D23880730)

[ghstack-poisoned]

Commit:
Update on "[FX][WIP] Mutable Graph APIs"


Introduced several `Graph` APIs:

`insert_node_before(self, n : Node, before : Node) -> Node:` -> insert node `n` directly before `before` in the topological ordering of nodes

`replace_all_uses_with(self, to_replace : Node, replace_with : Node)` -> Find all uses of the value produced by `to-replace` and swap out those references for the value produced by `replace_with`. Does not delete `to_replace`.

`erase_node(self, to_erase : Node)` -> Erases the node `to_erase` from the `Graph`. Throws an exception if there are still uses of that node in the graph.

`find_all_uses_of(self, to_find : Node) -> List[Node]` -> Find all nodes that use `to_find` in their args or kwargs

Also adds a context manager, `torch.fx.graph.WithInsertionPoint` that can be used to make node insertion happen *before* the passed-in node. For example:

```
graph = torch.fx.Graph()
x = torch.fx.Proxy(graph.placeholder('x'))
relu = torch.relu(x)

with torch.fx.graph.WithInsertPoint(relu.node):
    y = torch.neg(x)
    z = torch.tanh(y)

graph.output((relu.node, z.node))
```

Results in graph

```
graph(x):
    %neg_1 : [uses=1] = call_function[target=<built-in method neg of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    %tanh_1 : [uses=1] = call_function[target=<built-in method tanh of type object at 0x7fa87adb1500>](args = (%neg_1,), kwargs = {})
    %relu_1 : [uses=1] = call_function[target=<built-in method relu of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    return (%relu_1, %tanh_1)
```

Differential Revision: [D23880730](https://our.internmc.facebook.com/intern/diff/D23880730)

[ghstack-poisoned]

Commit:
Update on "[FX][WIP] Mutable Graph APIs"


Introduced several `Graph` APIs:

`insert_node_before(self, n : Node, before : Node) -> Node:` -> insert node `n` directly before `before` in the topological ordering of nodes

`replace_all_uses_with(self, to_replace : Node, replace_with : Node)` -> Find all uses of the value produced by `to-replace` and swap out those references for the value produced by `replace_with`. Does not delete `to_replace`.

`erase_node(self, to_erase : Node)` -> Erases the node `to_erase` from the `Graph`. Throws an exception if there are still uses of that node in the graph.

`find_all_uses_of(self, to_find : Node) -> List[Node]` -> Find all nodes that use `to_find` in their args or kwargs

Also adds a context manager, `torch.fx.graph.WithInsertionPoint` that can be used to make node insertion happen *before* the passed-in node. For example:

```
graph = torch.fx.Graph()
x = torch.fx.Proxy(graph.placeholder('x'))
relu = torch.relu(x)

with torch.fx.graph.WithInsertPoint(relu.node):
    y = torch.neg(x)
    z = torch.tanh(y)

graph.output((relu.node, z.node))
```

Results in graph

```
graph(x):
    %neg_1 : [uses=1] = call_function[target=<built-in method neg of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    %tanh_1 : [uses=1] = call_function[target=<built-in method tanh of type object at 0x7fa87adb1500>](args = (%neg_1,), kwargs = {})
    %relu_1 : [uses=1] = call_function[target=<built-in method relu of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    return (%relu_1, %tanh_1)
```

Differential Revision: [D23880730](https://our.internmc.facebook.com/intern/diff/D23880730)

[ghstack-poisoned]

Commit:
Update on "[FX][WIP] Mutable Graph APIs"


Introduced several `Graph` APIs:

`insert_node_before(self, n : Node, before : Node) -> Node:` -> insert node `n` directly before `before` in the topological ordering of nodes

`replace_all_uses_with(self, to_replace : Node, replace_with : Node)` -> Find all uses of the value produced by `to-replace` and swap out those references for the value produced by `replace_with`. Does not delete `to_replace`.

`erase_node(self, to_erase : Node)` -> Erases the node `to_erase` from the `Graph`. Throws an exception if there are still uses of that node in the graph.

`find_all_uses_of(self, to_find : Node) -> List[Node]` -> Find all nodes that use `to_find` in their args or kwargs

Also adds a context manager, `torch.fx.graph.WithInsertionPoint` that can be used to make node insertion happen *before* the passed-in node. For example:

```
graph = torch.fx.Graph()
x = torch.fx.Proxy(graph.placeholder('x'))
relu = torch.relu(x)

with torch.fx.graph.WithInsertPoint(relu.node):
    y = torch.neg(x)
    z = torch.tanh(y)

graph.output((relu.node, z.node))
```

Results in graph

```
graph(x):
    %neg_1 : [uses=1] = call_function[target=<built-in method neg of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    %tanh_1 : [uses=1] = call_function[target=<built-in method tanh of type object at 0x7fa87adb1500>](args = (%neg_1,), kwargs = {})
    %relu_1 : [uses=1] = call_function[target=<built-in method relu of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    return (%relu_1, %tanh_1)
```

Differential Revision: [D23880730](https://our.internmc.facebook.com/intern/diff/D23880730)

[ghstack-poisoned]

zdevito commented on torch/fx/graph.py:
```diff
@@ -65,10 +65,22 @@ def map_arg(a: Argument, fn: Callable[[Node], Argument]) -> Argument:
     else:
         return a
 
+class WithInsertPoint:
```
PyTorch convention is for resource guards to use snake case. `With` is already in the syntax for using it.  Probably should be more descriptive:
`with insert_before(n):`

zdevito commented on torch/fx/graph.py:
```diff
@@ -101,10 +113,78 @@ def create_node(self, op: str, target: Target,
+        self.insert_node_before(n=n, before=self._insert_point)
         return n
 
+    def insert_node_before(self, n : Node, before : Node) -> Node:
```
How do you get a node that is not already inserted into the graph? I can see `move_node_before` as being a public function, and I can see `create_node` as taking an optional before, but I don't see how this can be used.

zdevito commented on torch/fx/graph.py:
```diff
@@ -101,10 +113,78 @@ def create_node(self, op: str, target: Target,
+        Insert node `n` directly before `before` in the topological ordering of nodes
+        """
+        # TODO: computationally inefficient
+        before_idx = self._nodes.index(before)
```
In the common case of creating a graph, this hits the worst case N^2 behavior every time. While it is probably acceptable for having the mutation operators be inefficient to start, making _every_ graph creation N^2 is not.

jamesr66a replied:
How is this N^2? Inserting one-before-the-end should still be N, right? I.e. every iteration we would only need to move that one element (amortized for resizing)

Anyway, this implementation is going to totally change after i rebase past making `output` a node

zdevito commented on torch/fx/graph.py:
```diff
@@ -101,10 +113,78 @@ def create_node(self, op: str, target: Target,
+        Find all nodes that use `to_find` in their args or kwargs
+        """
+        # TODO: computationally inefficient
+        def find_use(arg : Any) -> bool:
```
The list of acceptable arguments should _only_ appear in a single place. Currently that is `map_arg`. Reproducing it here makes it really difficult to add other aggregates.

jamesr66a replied:
We'll probably need something like `reduce_arg` then

jamesr66a replied:
nvm, can hack this up with captures

zdevito commented on torch/fx/graph.py:
```diff
@@ -101,10 +113,78 @@ def create_node(self, op: str, target: Target,
+            assert isinstance(new_kwargs, dict)
+            node.kwargs = new_kwargs
+
+    def erase_node(self, to_erase : Node):
```
This fails to edit the `uses` attributes of its arguments.

jamesr66a replied:
A precondition for this function is that `to_erase` should already have `uses=0`. I don't think there's any other use editing that should be happening

zdevito commented on torch/fx/graph.py:
```diff
@@ -101,10 +113,78 @@ def create_node(self, op: str, target: Target,
+        for node in self._nodes:
+            map_arg(node.args, check_for_use)
+            map_arg(node.kwargs, check_for_use)
+        if self.result:
```
If this is the only place that has to be special because of uses in the return then I am not sure it is worth it

zdevito commented on torch/fx/graph.py:
```diff
@@ -101,10 +113,78 @@ def create_node(self, op: str, target: Target,
+                return n
+
+        # TODO: computationally inefficient
+        for node in self._nodes:
```
Doesn't edit the `uses` attribute on nodes.

zdevito commented on torch/fx/graph.py:
```diff
@@ -101,10 +113,78 @@ def create_node(self, op: str, target: Target,
+        self._nodes.insert(before_idx, n)
+        return n
+
+    def find_all_uses_of(self, to_find : Node) -> List[Node]:
```
`node.uses()` seems like the better place.

zdevito commented on torch/fx/graph.py:
```diff
@@ -101,10 +113,78 @@ def create_node(self, op: str, target: Target,
+
+        return [node for node in self._nodes if find_use(node.args) or find_use(node.kwargs)]
+
+    def replace_all_uses_with(self, to_replace : Node, replace_with : Node):
```
`node.replace_all_uses_with(other)`

zdevito reviewed (commented):
Thanks for putting this together. I put some comments inline. Overall I think we need to address:
* Common cases shouldn't be n^2. Things like iterating the nodes for every use search are bad because the O(N) operator is a interpreted python loop. The only acceptable non-linear things should be inserts into the node list or search the node list (C operations).
* We should resolve the behavior of placeholder and output nodes for consistency. I think the best thing to do is to make them both `Nodes` but for neither of them to appear in the list of nodes. This is similar to what ir.h arrived at and has good behavior: every use is a Node, every arg is a Node, but the special nodes that start/end the graph don't have to be handled in the if statements while iterating the node list.

Commit:
Update on "[FX][WIP] Mutable Graph APIs"


Introduced several `Graph` APIs:

`insert_node_before(self, n : Node, before : Node) -> Node:` -> insert node `n` directly before `before` in the topological ordering of nodes

`replace_all_uses_with(self, to_replace : Node, replace_with : Node)` -> Find all uses of the value produced by `to-replace` and swap out those references for the value produced by `replace_with`. Does not delete `to_replace`.

`erase_node(self, to_erase : Node)` -> Erases the node `to_erase` from the `Graph`. Throws an exception if there are still uses of that node in the graph.

`find_all_uses_of(self, to_find : Node) -> List[Node]` -> Find all nodes that use `to_find` in their args or kwargs

Also adds a context manager, `torch.fx.graph.WithInsertionPoint` that can be used to make node insertion happen *before* the passed-in node. For example:

```
graph = torch.fx.Graph()
x = torch.fx.Proxy(graph.placeholder('x'))
relu = torch.relu(x)

with torch.fx.graph.WithInsertPoint(relu.node):
    y = torch.neg(x)
    z = torch.tanh(y)

graph.output((relu.node, z.node))
```

Results in graph

```
graph(x):
    %neg_1 : [uses=1] = call_function[target=<built-in method neg of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    %tanh_1 : [uses=1] = call_function[target=<built-in method tanh of type object at 0x7fa87adb1500>](args = (%neg_1,), kwargs = {})
    %relu_1 : [uses=1] = call_function[target=<built-in method relu of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    return (%relu_1, %tanh_1)
```

Differential Revision: [D23880730](https://our.internmc.facebook.com/intern/diff/D23880730)

[ghstack-poisoned]

Commit:
Update on "[FX][WIP] Mutable Graph APIs"


Introduced several `Graph` APIs:

`insert_node_before(self, n : Node, before : Node) -> Node:` -> insert node `n` directly before `before` in the topological ordering of nodes

`replace_all_uses_with(self, to_replace : Node, replace_with : Node)` -> Find all uses of the value produced by `to-replace` and swap out those references for the value produced by `replace_with`. Does not delete `to_replace`.

`erase_node(self, to_erase : Node)` -> Erases the node `to_erase` from the `Graph`. Throws an exception if there are still uses of that node in the graph.

`find_all_uses_of(self, to_find : Node) -> List[Node]` -> Find all nodes that use `to_find` in their args or kwargs

Also adds a context manager, `torch.fx.graph.WithInsertionPoint` that can be used to make node insertion happen *before* the passed-in node. For example:

```
graph = torch.fx.Graph()
x = torch.fx.Proxy(graph.placeholder('x'))
relu = torch.relu(x)

with torch.fx.graph.WithInsertPoint(relu.node):
    y = torch.neg(x)
    z = torch.tanh(y)

graph.output((relu.node, z.node))
```

Results in graph

```
graph(x):
    %neg_1 : [uses=1] = call_function[target=<built-in method neg of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    %tanh_1 : [uses=1] = call_function[target=<built-in method tanh of type object at 0x7fa87adb1500>](args = (%neg_1,), kwargs = {})
    %relu_1 : [uses=1] = call_function[target=<built-in method relu of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    return (%relu_1, %tanh_1)
```

Differential Revision: [D23880730](https://our.internmc.facebook.com/intern/diff/D23880730)

[ghstack-poisoned]

Commit:
Update on "[FX][WIP] Mutable Graph APIs"


Introduced several `Graph` APIs:

`insert_node_before(self, n : Node, before : Node) -> Node:` -> insert node `n` directly before `before` in the topological ordering of nodes

`replace_all_uses_with(self, to_replace : Node, replace_with : Node)` -> Find all uses of the value produced by `to-replace` and swap out those references for the value produced by `replace_with`. Does not delete `to_replace`.

`erase_node(self, to_erase : Node)` -> Erases the node `to_erase` from the `Graph`. Throws an exception if there are still uses of that node in the graph.

`find_all_uses_of(self, to_find : Node) -> List[Node]` -> Find all nodes that use `to_find` in their args or kwargs

Also adds a context manager, `torch.fx.graph.WithInsertionPoint` that can be used to make node insertion happen *before* the passed-in node. For example:

```
graph = torch.fx.Graph()
x = torch.fx.Proxy(graph.placeholder('x'))
relu = torch.relu(x)

with torch.fx.graph.WithInsertPoint(relu.node):
    y = torch.neg(x)
    z = torch.tanh(y)

graph.output((relu.node, z.node))
```

Results in graph

```
graph(x):
    %neg_1 : [uses=1] = call_function[target=<built-in method neg of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    %tanh_1 : [uses=1] = call_function[target=<built-in method tanh of type object at 0x7fa87adb1500>](args = (%neg_1,), kwargs = {})
    %relu_1 : [uses=1] = call_function[target=<built-in method relu of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    return (%relu_1, %tanh_1)
```

Differential Revision: [D23880730](https://our.internmc.facebook.com/intern/diff/D23880730)

[ghstack-poisoned]

Commit:
Update on "[FX][WIP] Mutable Graph APIs"


Introduced several `Graph` APIs:

`insert_node_before(self, n : Node, before : Node) -> Node:` -> insert node `n` directly before `before` in the topological ordering of nodes

`replace_all_uses_with(self, to_replace : Node, replace_with : Node)` -> Find all uses of the value produced by `to-replace` and swap out those references for the value produced by `replace_with`. Does not delete `to_replace`.

`erase_node(self, to_erase : Node)` -> Erases the node `to_erase` from the `Graph`. Throws an exception if there are still uses of that node in the graph.

`find_all_uses_of(self, to_find : Node) -> List[Node]` -> Find all nodes that use `to_find` in their args or kwargs

Also adds a context manager, `torch.fx.graph.WithInsertionPoint` that can be used to make node insertion happen *before* the passed-in node. For example:

```
graph = torch.fx.Graph()
x = torch.fx.Proxy(graph.placeholder('x'))
relu = torch.relu(x)

with torch.fx.graph.WithInsertPoint(relu.node):
    y = torch.neg(x)
    z = torch.tanh(y)

graph.output((relu.node, z.node))
```

Results in graph

```
graph(x):
    %neg_1 : [uses=1] = call_function[target=<built-in method neg of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    %tanh_1 : [uses=1] = call_function[target=<built-in method tanh of type object at 0x7fa87adb1500>](args = (%neg_1,), kwargs = {})
    %relu_1 : [uses=1] = call_function[target=<built-in method relu of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    return (%relu_1, %tanh_1)
```

Differential Revision: [D23880730](https://our.internmc.facebook.com/intern/diff/D23880730)

[ghstack-poisoned]

Commit:
Update on "[FX][WIP] Mutable Graph APIs"


Introduced several `Graph` APIs:

`insert_node_before(self, n : Node, before : Node) -> Node:` -> insert node `n` directly before `before` in the topological ordering of nodes

`replace_all_uses_with(self, to_replace : Node, replace_with : Node)` -> Find all uses of the value produced by `to-replace` and swap out those references for the value produced by `replace_with`. Does not delete `to_replace`.

`erase_node(self, to_erase : Node)` -> Erases the node `to_erase` from the `Graph`. Throws an exception if there are still uses of that node in the graph.

`find_all_uses_of(self, to_find : Node) -> List[Node]` -> Find all nodes that use `to_find` in their args or kwargs

Also adds a context manager, `torch.fx.graph.WithInsertionPoint` that can be used to make node insertion happen *before* the passed-in node. For example:

```
graph = torch.fx.Graph()
x = torch.fx.Proxy(graph.placeholder('x'))
relu = torch.relu(x)

with torch.fx.graph.WithInsertPoint(relu.node):
    y = torch.neg(x)
    z = torch.tanh(y)

graph.output((relu.node, z.node))
```

Results in graph

```
graph(x):
    %neg_1 : [uses=1] = call_function[target=<built-in method neg of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    %tanh_1 : [uses=1] = call_function[target=<built-in method tanh of type object at 0x7fa87adb1500>](args = (%neg_1,), kwargs = {})
    %relu_1 : [uses=1] = call_function[target=<built-in method relu of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    return (%relu_1, %tanh_1)
```

Differential Revision: [D23880730](https://our.internmc.facebook.com/intern/diff/D23880730)

[ghstack-poisoned]

Commit:
Update on "[FX][WIP] Mutable Graph APIs"


Introduced several `Graph` APIs:

`insert_node_before(self, n : Node, before : Node) -> Node:` -> insert node `n` directly before `before` in the topological ordering of nodes

`replace_all_uses_with(self, to_replace : Node, replace_with : Node)` -> Find all uses of the value produced by `to-replace` and swap out those references for the value produced by `replace_with`. Does not delete `to_replace`.

`erase_node(self, to_erase : Node)` -> Erases the node `to_erase` from the `Graph`. Throws an exception if there are still uses of that node in the graph.

`find_all_uses_of(self, to_find : Node) -> List[Node]` -> Find all nodes that use `to_find` in their args or kwargs

Also adds a context manager, `torch.fx.graph.WithInsertionPoint` that can be used to make node insertion happen *before* the passed-in node. For example:

```
graph = torch.fx.Graph()
x = torch.fx.Proxy(graph.placeholder('x'))
relu = torch.relu(x)

with torch.fx.graph.WithInsertPoint(relu.node):
    y = torch.neg(x)
    z = torch.tanh(y)

graph.output((relu.node, z.node))
```

Results in graph

```
graph(x):
    %neg_1 : [uses=1] = call_function[target=<built-in method neg of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    %tanh_1 : [uses=1] = call_function[target=<built-in method tanh of type object at 0x7fa87adb1500>](args = (%neg_1,), kwargs = {})
    %relu_1 : [uses=1] = call_function[target=<built-in method relu of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    return (%relu_1, %tanh_1)
```

Differential Revision: [D23880730](https://our.internmc.facebook.com/intern/diff/D23880730)

[ghstack-poisoned]

Commit:
Update on "[FX][WIP] Mutable Graph APIs"


Introduced several `Graph` APIs:

`insert_node_before(self, n : Node, before : Node) -> Node:` -> insert node `n` directly before `before` in the topological ordering of nodes

`replace_all_uses_with(self, to_replace : Node, replace_with : Node)` -> Find all uses of the value produced by `to-replace` and swap out those references for the value produced by `replace_with`. Does not delete `to_replace`.

`erase_node(self, to_erase : Node)` -> Erases the node `to_erase` from the `Graph`. Throws an exception if there are still uses of that node in the graph.

`find_all_uses_of(self, to_find : Node) -> List[Node]` -> Find all nodes that use `to_find` in their args or kwargs

Also adds a context manager, `torch.fx.graph.WithInsertionPoint` that can be used to make node insertion happen *before* the passed-in node. For example:

```
graph = torch.fx.Graph()
x = torch.fx.Proxy(graph.placeholder('x'))
relu = torch.relu(x)

with torch.fx.graph.WithInsertPoint(relu.node):
    y = torch.neg(x)
    z = torch.tanh(y)

graph.output((relu.node, z.node))
```

Results in graph

```
graph(x):
    %neg_1 : [uses=1] = call_function[target=<built-in method neg of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    %tanh_1 : [uses=1] = call_function[target=<built-in method tanh of type object at 0x7fa87adb1500>](args = (%neg_1,), kwargs = {})
    %relu_1 : [uses=1] = call_function[target=<built-in method relu of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    return (%relu_1, %tanh_1)
```

Differential Revision: [D23880730](https://our.internmc.facebook.com/intern/diff/D23880730)

[ghstack-poisoned]

Commit:
Update on "[FX][WIP] Mutable Graph APIs"


Introduced several `Graph` APIs:

`insert_node_before(self, n : Node, before : Node) -> Node:` -> insert node `n` directly before `before` in the topological ordering of nodes

`replace_all_uses_with(self, to_replace : Node, replace_with : Node)` -> Find all uses of the value produced by `to-replace` and swap out those references for the value produced by `replace_with`. Does not delete `to_replace`.

`erase_node(self, to_erase : Node)` -> Erases the node `to_erase` from the `Graph`. Throws an exception if there are still uses of that node in the graph.

`find_all_uses_of(self, to_find : Node) -> List[Node]` -> Find all nodes that use `to_find` in their args or kwargs

Also adds a context manager, `torch.fx.graph.WithInsertionPoint` that can be used to make node insertion happen *before* the passed-in node. For example:

```
graph = torch.fx.Graph()
x = torch.fx.Proxy(graph.placeholder('x'))
relu = torch.relu(x)

with torch.fx.graph.WithInsertPoint(relu.node):
    y = torch.neg(x)
    z = torch.tanh(y)

graph.output((relu.node, z.node))
```

Results in graph

```
graph(x):
    %neg_1 : [uses=1] = call_function[target=<built-in method neg of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    %tanh_1 : [uses=1] = call_function[target=<built-in method tanh of type object at 0x7fa87adb1500>](args = (%neg_1,), kwargs = {})
    %relu_1 : [uses=1] = call_function[target=<built-in method relu of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    return (%relu_1, %tanh_1)
```

Differential Revision: [D23880730](https://our.internmc.facebook.com/intern/diff/D23880730)

[ghstack-poisoned]

Commit:
Update on "[FX][WIP] Mutable Graph APIs"


Introduced several `Graph` APIs:

`insert_node_before(self, n : Node, before : Node) -> Node:` -> insert node `n` directly before `before` in the topological ordering of nodes

`replace_all_uses_with(self, to_replace : Node, replace_with : Node)` -> Find all uses of the value produced by `to-replace` and swap out those references for the value produced by `replace_with`. Does not delete `to_replace`.

`erase_node(self, to_erase : Node)` -> Erases the node `to_erase` from the `Graph`. Throws an exception if there are still uses of that node in the graph.

`find_all_uses_of(self, to_find : Node) -> List[Node]` -> Find all nodes that use `to_find` in their args or kwargs

Also adds a context manager, `torch.fx.graph.WithInsertionPoint` that can be used to make node insertion happen *before* the passed-in node. For example:

```
graph = torch.fx.Graph()
x = torch.fx.Proxy(graph.placeholder('x'))
relu = torch.relu(x)

with torch.fx.graph.WithInsertPoint(relu.node):
    y = torch.neg(x)
    z = torch.tanh(y)

graph.output((relu.node, z.node))
```

Results in graph

```
graph(x):
    %neg_1 : [uses=1] = call_function[target=<built-in method neg of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    %tanh_1 : [uses=1] = call_function[target=<built-in method tanh of type object at 0x7fa87adb1500>](args = (%neg_1,), kwargs = {})
    %relu_1 : [uses=1] = call_function[target=<built-in method relu of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    return (%relu_1, %tanh_1)
```

Differential Revision: [D23880730](https://our.internmc.facebook.com/intern/diff/D23880730)

[ghstack-poisoned]

Commit:
Update on "[FX][WIP] Mutable Graph APIs"


Introduced several `Graph` APIs:

`insert_node_before(self, n : Node, before : Node) -> Node:` -> insert node `n` directly before `before` in the topological ordering of nodes

`replace_all_uses_with(self, to_replace : Node, replace_with : Node)` -> Find all uses of the value produced by `to-replace` and swap out those references for the value produced by `replace_with`. Does not delete `to_replace`.

`erase_node(self, to_erase : Node)` -> Erases the node `to_erase` from the `Graph`. Throws an exception if there are still uses of that node in the graph.

`find_all_uses_of(self, to_find : Node) -> List[Node]` -> Find all nodes that use `to_find` in their args or kwargs

Also adds a context manager, `torch.fx.graph.WithInsertionPoint` that can be used to make node insertion happen *before* the passed-in node. For example:

```
graph = torch.fx.Graph()
x = torch.fx.Proxy(graph.placeholder('x'))
relu = torch.relu(x)

with torch.fx.graph.WithInsertPoint(relu.node):
    y = torch.neg(x)
    z = torch.tanh(y)

graph.output((relu.node, z.node))
```

Results in graph

```
graph(x):
    %neg_1 : [uses=1] = call_function[target=<built-in method neg of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    %tanh_1 : [uses=1] = call_function[target=<built-in method tanh of type object at 0x7fa87adb1500>](args = (%neg_1,), kwargs = {})
    %relu_1 : [uses=1] = call_function[target=<built-in method relu of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    return (%relu_1, %tanh_1)
```

Differential Revision: [D23880730](https://our.internmc.facebook.com/intern/diff/D23880730)

[ghstack-poisoned]

Commit:
Update on "[FX][WIP] Mutable Graph APIs"


Introduced several `Graph` APIs:

`insert_node_before(self, n : Node, before : Node) -> Node:` -> insert node `n` directly before `before` in the topological ordering of nodes

`replace_all_uses_with(self, to_replace : Node, replace_with : Node)` -> Find all uses of the value produced by `to-replace` and swap out those references for the value produced by `replace_with`. Does not delete `to_replace`.

`erase_node(self, to_erase : Node)` -> Erases the node `to_erase` from the `Graph`. Throws an exception if there are still uses of that node in the graph.

`find_all_uses_of(self, to_find : Node) -> List[Node]` -> Find all nodes that use `to_find` in their args or kwargs

Also adds a context manager, `torch.fx.graph.WithInsertionPoint` that can be used to make node insertion happen *before* the passed-in node. For example:

```
graph = torch.fx.Graph()
x = torch.fx.Proxy(graph.placeholder('x'))
relu = torch.relu(x)

with torch.fx.graph.WithInsertPoint(relu.node):
    y = torch.neg(x)
    z = torch.tanh(y)

graph.output((relu.node, z.node))
```

Results in graph

```
graph(x):
    %neg_1 : [uses=1] = call_function[target=<built-in method neg of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    %tanh_1 : [uses=1] = call_function[target=<built-in method tanh of type object at 0x7fa87adb1500>](args = (%neg_1,), kwargs = {})
    %relu_1 : [uses=1] = call_function[target=<built-in method relu of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    return (%relu_1, %tanh_1)
```

Differential Revision: [D23880730](https://our.internmc.facebook.com/intern/diff/D23880730)

[ghstack-poisoned]

Commit:
Update on "[FX][WIP] Mutable Graph APIs"


Introduced several `Graph` APIs:

`insert_node_before(self, n : Node, before : Node) -> Node:` -> insert node `n` directly before `before` in the topological ordering of nodes

`replace_all_uses_with(self, to_replace : Node, replace_with : Node)` -> Find all uses of the value produced by `to-replace` and swap out those references for the value produced by `replace_with`. Does not delete `to_replace`.

`erase_node(self, to_erase : Node)` -> Erases the node `to_erase` from the `Graph`. Throws an exception if there are still uses of that node in the graph.

`find_all_uses_of(self, to_find : Node) -> List[Node]` -> Find all nodes that use `to_find` in their args or kwargs

Also adds a context manager, `torch.fx.graph.WithInsertionPoint` that can be used to make node insertion happen *before* the passed-in node. For example:

```
graph = torch.fx.Graph()
x = torch.fx.Proxy(graph.placeholder('x'))
relu = torch.relu(x)

with torch.fx.graph.WithInsertPoint(relu.node):
    y = torch.neg(x)
    z = torch.tanh(y)

graph.output((relu.node, z.node))
```

Results in graph

```
graph(x):
    %neg_1 : [uses=1] = call_function[target=<built-in method neg of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    %tanh_1 : [uses=1] = call_function[target=<built-in method tanh of type object at 0x7fa87adb1500>](args = (%neg_1,), kwargs = {})
    %relu_1 : [uses=1] = call_function[target=<built-in method relu of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    return (%relu_1, %tanh_1)
```

Differential Revision: [D23880730](https://our.internmc.facebook.com/intern/diff/D23880730)

[ghstack-poisoned]

Commit:
Update on "[FX][WIP] Mutable Graph APIs"


Introduced several `Graph` APIs:

`insert_node_before(self, n : Node, before : Node) -> Node:` -> insert node `n` directly before `before` in the topological ordering of nodes

`replace_all_uses_with(self, to_replace : Node, replace_with : Node)` -> Find all uses of the value produced by `to-replace` and swap out those references for the value produced by `replace_with`. Does not delete `to_replace`.

`erase_node(self, to_erase : Node)` -> Erases the node `to_erase` from the `Graph`. Throws an exception if there are still uses of that node in the graph.

`find_all_uses_of(self, to_find : Node) -> List[Node]` -> Find all nodes that use `to_find` in their args or kwargs

Also adds a context manager, `torch.fx.graph.WithInsertionPoint` that can be used to make node insertion happen *before* the passed-in node. For example:

```
graph = torch.fx.Graph()
x = torch.fx.Proxy(graph.placeholder('x'))
relu = torch.relu(x)

with torch.fx.graph.WithInsertPoint(relu.node):
    y = torch.neg(x)
    z = torch.tanh(y)

graph.output((relu.node, z.node))
```

Results in graph

```
graph(x):
    %neg_1 : [uses=1] = call_function[target=<built-in method neg of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    %tanh_1 : [uses=1] = call_function[target=<built-in method tanh of type object at 0x7fa87adb1500>](args = (%neg_1,), kwargs = {})
    %relu_1 : [uses=1] = call_function[target=<built-in method relu of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    return (%relu_1, %tanh_1)
```

Differential Revision: [D23880730](https://our.internmc.facebook.com/intern/diff/D23880730)

[ghstack-poisoned]

Commit:
Update on "[FX][WIP] Mutable Graph APIs"


Introduced several `Graph` APIs:

`insert_node_before(self, n : Node, before : Node) -> Node:` -> insert node `n` directly before `before` in the topological ordering of nodes

`replace_all_uses_with(self, to_replace : Node, replace_with : Node)` -> Find all uses of the value produced by `to-replace` and swap out those references for the value produced by `replace_with`. Does not delete `to_replace`.

`erase_node(self, to_erase : Node)` -> Erases the node `to_erase` from the `Graph`. Throws an exception if there are still uses of that node in the graph.

`find_all_uses_of(self, to_find : Node) -> List[Node]` -> Find all nodes that use `to_find` in their args or kwargs

Also adds a context manager, `torch.fx.graph.WithInsertionPoint` that can be used to make node insertion happen *before* the passed-in node. For example:

```
graph = torch.fx.Graph()
x = torch.fx.Proxy(graph.placeholder('x'))
relu = torch.relu(x)

with torch.fx.graph.WithInsertPoint(relu.node):
    y = torch.neg(x)
    z = torch.tanh(y)

graph.output((relu.node, z.node))
```

Results in graph

```
graph(x):
    %neg_1 : [uses=1] = call_function[target=<built-in method neg of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    %tanh_1 : [uses=1] = call_function[target=<built-in method tanh of type object at 0x7fa87adb1500>](args = (%neg_1,), kwargs = {})
    %relu_1 : [uses=1] = call_function[target=<built-in method relu of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    return (%relu_1, %tanh_1)
```

Differential Revision: [D23880730](https://our.internmc.facebook.com/intern/diff/D23880730)

[ghstack-poisoned]

Commit:
Update on "[FX][WIP] Mutable Graph APIs"


Introduced several `Graph` APIs:

`insert_node_before(self, n : Node, before : Node) -> Node:` -> insert node `n` directly before `before` in the topological ordering of nodes

`replace_all_uses_with(self, to_replace : Node, replace_with : Node)` -> Find all uses of the value produced by `to-replace` and swap out those references for the value produced by `replace_with`. Does not delete `to_replace`.

`erase_node(self, to_erase : Node)` -> Erases the node `to_erase` from the `Graph`. Throws an exception if there are still uses of that node in the graph.

`find_all_uses_of(self, to_find : Node) -> List[Node]` -> Find all nodes that use `to_find` in their args or kwargs

Also adds a context manager, `torch.fx.graph.WithInsertionPoint` that can be used to make node insertion happen *before* the passed-in node. For example:

```
graph = torch.fx.Graph()
x = torch.fx.Proxy(graph.placeholder('x'))
relu = torch.relu(x)

with torch.fx.graph.WithInsertPoint(relu.node):
    y = torch.neg(x)
    z = torch.tanh(y)

graph.output((relu.node, z.node))
```

Results in graph

```
graph(x):
    %neg_1 : [uses=1] = call_function[target=<built-in method neg of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    %tanh_1 : [uses=1] = call_function[target=<built-in method tanh of type object at 0x7fa87adb1500>](args = (%neg_1,), kwargs = {})
    %relu_1 : [uses=1] = call_function[target=<built-in method relu of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    return (%relu_1, %tanh_1)
```

Differential Revision: [D23880730](https://our.internmc.facebook.com/intern/diff/D23880730)

[ghstack-poisoned]

Commit:
Update on "[FX][WIP] Mutable Graph APIs"


Introduced several `Graph` APIs:

`insert_node_before(self, n : Node, before : Node) -> Node:` -> insert node `n` directly before `before` in the topological ordering of nodes

`replace_all_uses_with(self, to_replace : Node, replace_with : Node)` -> Find all uses of the value produced by `to-replace` and swap out those references for the value produced by `replace_with`. Does not delete `to_replace`.

`erase_node(self, to_erase : Node)` -> Erases the node `to_erase` from the `Graph`. Throws an exception if there are still uses of that node in the graph.

`find_all_uses_of(self, to_find : Node) -> List[Node]` -> Find all nodes that use `to_find` in their args or kwargs

Also adds a context manager, `torch.fx.graph.WithInsertionPoint` that can be used to make node insertion happen *before* the passed-in node. For example:

```
graph = torch.fx.Graph()
x = torch.fx.Proxy(graph.placeholder('x'))
relu = torch.relu(x)

with torch.fx.graph.WithInsertPoint(relu.node):
    y = torch.neg(x)
    z = torch.tanh(y)

graph.output((relu.node, z.node))
```

Results in graph

```
graph(x):
    %neg_1 : [uses=1] = call_function[target=<built-in method neg of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    %tanh_1 : [uses=1] = call_function[target=<built-in method tanh of type object at 0x7fa87adb1500>](args = (%neg_1,), kwargs = {})
    %relu_1 : [uses=1] = call_function[target=<built-in method relu of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    return (%relu_1, %tanh_1)
```

Differential Revision: [D23880730](https://our.internmc.facebook.com/intern/diff/D23880730)

[ghstack-poisoned]

Commit:
Update on "[FX][WIP] Mutable Graph APIs"


Introduced several `Graph` APIs:

`insert_node_before(self, n : Node, before : Node) -> Node:` -> insert node `n` directly before `before` in the topological ordering of nodes

`replace_all_uses_with(self, to_replace : Node, replace_with : Node)` -> Find all uses of the value produced by `to-replace` and swap out those references for the value produced by `replace_with`. Does not delete `to_replace`.

`erase_node(self, to_erase : Node)` -> Erases the node `to_erase` from the `Graph`. Throws an exception if there are still uses of that node in the graph.

`find_all_uses_of(self, to_find : Node) -> List[Node]` -> Find all nodes that use `to_find` in their args or kwargs

Also adds a context manager, `torch.fx.graph.WithInsertionPoint` that can be used to make node insertion happen *before* the passed-in node. For example:

```
graph = torch.fx.Graph()
x = torch.fx.Proxy(graph.placeholder('x'))
relu = torch.relu(x)

with torch.fx.graph.WithInsertPoint(relu.node):
    y = torch.neg(x)
    z = torch.tanh(y)

graph.output((relu.node, z.node))
```

Results in graph

```
graph(x):
    %neg_1 : [uses=1] = call_function[target=<built-in method neg of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    %tanh_1 : [uses=1] = call_function[target=<built-in method tanh of type object at 0x7fa87adb1500>](args = (%neg_1,), kwargs = {})
    %relu_1 : [uses=1] = call_function[target=<built-in method relu of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    return (%relu_1, %tanh_1)
```

Differential Revision: [D23880730](https://our.internmc.facebook.com/intern/diff/D23880730)

[ghstack-poisoned]

Commit:
Update on "[FX][WIP] Mutable Graph APIs"


Introduced several `Graph` APIs:

`insert_node_before(self, n : Node, before : Node) -> Node:` -> insert node `n` directly before `before` in the topological ordering of nodes

`replace_all_uses_with(self, to_replace : Node, replace_with : Node)` -> Find all uses of the value produced by `to-replace` and swap out those references for the value produced by `replace_with`. Does not delete `to_replace`.

`erase_node(self, to_erase : Node)` -> Erases the node `to_erase` from the `Graph`. Throws an exception if there are still uses of that node in the graph.

`find_all_uses_of(self, to_find : Node) -> List[Node]` -> Find all nodes that use `to_find` in their args or kwargs

Also adds a context manager, `torch.fx.graph.WithInsertionPoint` that can be used to make node insertion happen *before* the passed-in node. For example:

```
graph = torch.fx.Graph()
x = torch.fx.Proxy(graph.placeholder('x'))
relu = torch.relu(x)

with torch.fx.graph.WithInsertPoint(relu.node):
    y = torch.neg(x)
    z = torch.tanh(y)

graph.output((relu.node, z.node))
```

Results in graph

```
graph(x):
    %neg_1 : [uses=1] = call_function[target=<built-in method neg of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    %tanh_1 : [uses=1] = call_function[target=<built-in method tanh of type object at 0x7fa87adb1500>](args = (%neg_1,), kwargs = {})
    %relu_1 : [uses=1] = call_function[target=<built-in method relu of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    return (%relu_1, %tanh_1)
```

Differential Revision: [D23880730](https://our.internmc.facebook.com/intern/diff/D23880730)

[ghstack-poisoned]

Commit:
Update on "[FX][WIP] Mutable Graph APIs"


Introduced several `Graph` APIs:

`insert_node_before(self, n : Node, before : Node) -> Node:` -> insert node `n` directly before `before` in the topological ordering of nodes

`replace_all_uses_with(self, to_replace : Node, replace_with : Node)` -> Find all uses of the value produced by `to-replace` and swap out those references for the value produced by `replace_with`. Does not delete `to_replace`.

`erase_node(self, to_erase : Node)` -> Erases the node `to_erase` from the `Graph`. Throws an exception if there are still uses of that node in the graph.

`find_all_uses_of(self, to_find : Node) -> List[Node]` -> Find all nodes that use `to_find` in their args or kwargs

Also adds a context manager, `torch.fx.graph.WithInsertionPoint` that can be used to make node insertion happen *before* the passed-in node. For example:

```
graph = torch.fx.Graph()
x = torch.fx.Proxy(graph.placeholder('x'))
relu = torch.relu(x)

with torch.fx.graph.WithInsertPoint(relu.node):
    y = torch.neg(x)
    z = torch.tanh(y)

graph.output((relu.node, z.node))
```

Results in graph

```
graph(x):
    %neg_1 : [uses=1] = call_function[target=<built-in method neg of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    %tanh_1 : [uses=1] = call_function[target=<built-in method tanh of type object at 0x7fa87adb1500>](args = (%neg_1,), kwargs = {})
    %relu_1 : [uses=1] = call_function[target=<built-in method relu of type object at 0x7fa87adb1500>](args = (%x,), kwargs = {})
    return (%relu_1, %tanh_1)
```

Differential Revision: [D23880730](https://our.internmc.facebook.com/intern/diff/D23880730)

[ghstack-poisoned]

Commit:
Update on "[FX][WIP] Mutable Graph APIs"


Introduced several `Graph` APIs:

`Graph.move_node_before(self, to_move : Node, before : Node)` -> Move node `to_move` before `before` in the Graph. Both `Node` arguments must be present in this graph.

`Graph.erase_node(self, to_erase : Node)` -> Erases the node `to_erase` from the `Graph`. Throws an exception if there are still uses of that node in the `Graph`.

`Node.find_uses(self) -> List['Node']` -> Find all nodes that use the value produced by `self`. The complexity of this function is linear in the number of nodes * number of arguments to each node.
Note that len(find_uses()) is not necessarily equal to attribute `uses`. This node could be used multiple times in the same `Node`. In that case, the user node would appear once in the return value here, but `uses` would account for the total number of times this Node is used by the user node. e.g. a node for `x + x` would have two uses for the `x` node, but the `x + x` node would appear once in the return from `find_uses`

`Node.replace_all_uses_with(self, replace_with : 'Node') -> List['Node']` -> Replace all uses of `self` in the Graph with the Node `replace_with`. Returns the list of nodes on which this change was made.

Also adds a context manager, `torch.fx.graph.insert_before` that can be used to make node insertion happen *before* the passed-in node. For example:

```
graph = torch.fx.Graph()
x = torch.fx.Proxy(graph.placeholder('x'))
relu = torch.relu(x)

with torch.fx.graph.insert_before(relu.node):
    y = torch.neg(x)
    z = torch.tanh(y)

graph.output((relu.node, z.node))
```

Results in graph

```
graph(x):
    %neg_1 : [uses=1] = call_function[target=<built-in method neg of type object at 0x7f6d19a3d6a0>](args = (%x,), kwargs = {})
    %tanh_1 : [uses=2] = call_function[target=<built-in method tanh of type object at 0x7f6d19a3d6a0>](args = (%neg_1,), kwargs = {})
    %relu_1 : [uses=2] = call_function[target=<built-in method relu of type object at 0x7f6d19a3d6a0>](args = (%x,), kwargs = {})
    return (relu_1, tanh_1)
```

Differential Revision: [D23880730](https://our.internmc.facebook.com/intern/diff/D23880730)

[ghstack-poisoned]

Commit:
Update on "[FX][WIP] Mutable Graph APIs"


Introduced several `Graph` APIs:

`Graph.move_node_before(self, to_move : Node, before : Node)` -> Move node `to_move` before `before` in the Graph. Both `Node` arguments must be present in this graph.

`Graph.erase_node(self, to_erase : Node)` -> Erases the node `to_erase` from the `Graph`. Throws an exception if there are still uses of that node in the `Graph`.

`Node.find_uses(self) -> List['Node']` -> Find all nodes that use the value produced by `self`. The complexity of this function is linear in the number of nodes * number of arguments to each node.
Note that len(find_uses()) is not necessarily equal to attribute `uses`. This node could be used multiple times in the same `Node`. In that case, the user node would appear once in the return value here, but `uses` would account for the total number of times this Node is used by the user node. e.g. a node for `x + x` would have two uses for the `x` node, but the `x + x` node would appear once in the return from `find_uses`

`Node.replace_all_uses_with(self, replace_with : 'Node') -> List['Node']` -> Replace all uses of `self` in the Graph with the Node `replace_with`. Returns the list of nodes on which this change was made.

Also adds a context manager, `torch.fx.graph.insert_before` that can be used to make node insertion happen *before* the passed-in node. For example:

```
graph = torch.fx.Graph()
x = torch.fx.Proxy(graph.placeholder('x'))
relu = torch.relu(x)

with torch.fx.graph.insert_before(relu.node):
    y = torch.neg(x)
    z = torch.tanh(y)

graph.output((relu.node, z.node))
```

Results in graph

```
graph(x):
    %neg_1 : [uses=1] = call_function[target=<built-in method neg of type object at 0x7f6d19a3d6a0>](args = (%x,), kwargs = {})
    %tanh_1 : [uses=2] = call_function[target=<built-in method tanh of type object at 0x7f6d19a3d6a0>](args = (%neg_1,), kwargs = {})
    %relu_1 : [uses=2] = call_function[target=<built-in method relu of type object at 0x7f6d19a3d6a0>](args = (%x,), kwargs = {})
    return (relu_1, tanh_1)
```

Differential Revision: [D23880730](https://our.internmc.facebook.com/intern/diff/D23880730)

[ghstack-poisoned]

Commit:
Update on "[FX] Mutable Graph APIs"


Introduced several `Graph` APIs:

`Graph.move_node_before(self, to_move : Node, before : Node)` -> Move node `to_move` before `before` in the Graph. Both `Node` arguments must be present in this graph.

`Graph.erase_node(self, to_erase : Node)` -> Erases the node `to_erase` from the `Graph`. Throws an exception if there are still uses of that node in the `Graph`.

`Node.find_uses(self) -> List['Node']` -> Find all nodes that use the value produced by `self`. The complexity of this function is linear in the number of nodes * number of arguments to each node.
Note that len(find_uses()) is not necessarily equal to attribute `uses`. This node could be used multiple times in the same `Node`. In that case, the user node would appear once in the return value here, but `uses` would account for the total number of times this Node is used by the user node. e.g. a node for `x + x` would have two uses for the `x` node, but the `x + x` node would appear once in the return from `find_uses`

`Node.replace_all_uses_with(self, replace_with : 'Node') -> List['Node']` -> Replace all uses of `self` in the Graph with the Node `replace_with`. Returns the list of nodes on which this change was made.

Also adds a context manager, `torch.fx.graph.insert_before` that can be used to make node insertion happen *before* the passed-in node. For example:

```
graph = torch.fx.Graph()
x = torch.fx.Proxy(graph.placeholder('x'))
relu = torch.relu(x)

with torch.fx.graph.insert_before(relu.node):
    y = torch.neg(x)
    z = torch.tanh(y)

graph.output((relu.node, z.node))
```

Results in graph

```
graph(x):
    %neg_1 : [uses=1] = call_function[target=<built-in method neg of type object at 0x7f6d19a3d6a0>](args = (%x,), kwargs = {})
    %tanh_1 : [uses=2] = call_function[target=<built-in method tanh of type object at 0x7f6d19a3d6a0>](args = (%neg_1,), kwargs = {})
    %relu_1 : [uses=2] = call_function[target=<built-in method relu of type object at 0x7f6d19a3d6a0>](args = (%x,), kwargs = {})
    return (relu_1, tanh_1)
```

Differential Revision: [D23880730](https://our.internmc.facebook.com/intern/diff/D23880730)

[ghstack-poisoned]

Commit:
Update on "[FX] Mutable Graph APIs"


Introduced several `Graph` APIs:

`Graph.move_node_before(self, to_move : Node, before : Node)` -> Move node `to_move` before `before` in the Graph. Both `Node` arguments must be present in this graph.

`Graph.erase_node(self, to_erase : Node)` -> Erases the node `to_erase` from the `Graph`. Throws an exception if there are still uses of that node in the `Graph`.

`Node.find_uses(self) -> List['Node']` -> Find all nodes that use the value produced by `self`. The complexity of this function is linear in the number of nodes * number of arguments to each node.
Note that len(find_uses()) is not necessarily equal to attribute `uses`. This node could be used multiple times in the same `Node`. In that case, the user node would appear once in the return value here, but `uses` would account for the total number of times this Node is used by the user node. e.g. a node for `x + x` would have two uses for the `x` node, but the `x + x` node would appear once in the return from `find_uses`

`Node.replace_all_uses_with(self, replace_with : 'Node') -> List['Node']` -> Replace all uses of `self` in the Graph with the Node `replace_with`. Returns the list of nodes on which this change was made.

Also adds a context manager, `torch.fx.graph.insert_before` that can be used to make node insertion happen *before* the passed-in node. For example:

```
graph = torch.fx.Graph()
x = torch.fx.Proxy(graph.placeholder('x'))
relu = torch.relu(x)

with torch.fx.graph.insert_before(relu.node):
    y = torch.neg(x)
    z = torch.tanh(y)

graph.output((relu.node, z.node))
```

Results in graph

```
graph(x):
    %neg_1 : [uses=1] = call_function[target=<built-in method neg of type object at 0x7f6d19a3d6a0>](args = (%x,), kwargs = {})
    %tanh_1 : [uses=2] = call_function[target=<built-in method tanh of type object at 0x7f6d19a3d6a0>](args = (%neg_1,), kwargs = {})
    %relu_1 : [uses=2] = call_function[target=<built-in method relu of type object at 0x7f6d19a3d6a0>](args = (%x,), kwargs = {})
    return (relu_1, tanh_1)
```

Differential Revision: [D23880730](https://our.internmc.facebook.com/intern/diff/D23880730)

[ghstack-poisoned]

Commit:
Update on "[FX] Mutable Graph APIs"


Introduced several `Graph` APIs:

`Graph.move_node_before(self, to_move : Node, before : Node)` -> Move node `to_move` before `before` in the Graph. Both `Node` arguments must be present in this graph.

`Graph.erase_node(self, to_erase : Node)` -> Erases the node `to_erase` from the `Graph`. Throws an exception if there are still uses of that node in the `Graph`.

`Node.find_uses(self) -> List['Node']` -> Find all nodes that use the value produced by `self`. The complexity of this function is linear in the number of nodes * number of arguments to each node.
Note that len(find_uses()) is not necessarily equal to attribute `uses`. This node could be used multiple times in the same `Node`. In that case, the user node would appear once in the return value here, but `uses` would account for the total number of times this Node is used by the user node. e.g. a node for `x + x` would have two uses for the `x` node, but the `x + x` node would appear once in the return from `find_uses`

`Node.replace_all_uses_with(self, replace_with : 'Node') -> List['Node']` -> Replace all uses of `self` in the Graph with the Node `replace_with`. Returns the list of nodes on which this change was made.

Also adds a context manager, `torch.fx.graph.insert_before` that can be used to make node insertion happen *before* the passed-in node. For example:

```
graph = torch.fx.Graph()
x = torch.fx.Proxy(graph.placeholder('x'))
relu = torch.relu(x)

with torch.fx.graph.insert_before(relu.node):
    y = torch.neg(x)
    z = torch.tanh(y)

graph.output((relu.node, z.node))
```

Results in graph

```
graph(x):
    %neg_1 : [uses=1] = call_function[target=<built-in method neg of type object at 0x7f6d19a3d6a0>](args = (%x,), kwargs = {})
    %tanh_1 : [uses=2] = call_function[target=<built-in method tanh of type object at 0x7f6d19a3d6a0>](args = (%neg_1,), kwargs = {})
    %relu_1 : [uses=2] = call_function[target=<built-in method relu of type object at 0x7f6d19a3d6a0>](args = (%x,), kwargs = {})
    return (relu_1, tanh_1)
```

Differential Revision: [D23880730](https://our.internmc.facebook.com/intern/diff/D23880730)

[ghstack-poisoned]
